### ISSUE TITLE
[Question] PVCs takes a long time (or fails) to be attached due to fsgroup change

### ISSUE THREAD

**Issue by nicolamarella** [longhorn/longhorn#1221]
We are currently using Kubernetes v1.17.4 and Longhorn 0.8.0, before upgrading to 0.8.0 the creation and attachment of PVs was quite slow, then we upgraded to 0.8.0 and we've noticed a sensible improvement in that regard. 

Lately, although, we've been experiencing the same, if not worse, time to get PVs attached and it's been very hard to us to find the root cause. A temporary solution for us was to cordone one by one each node and then restarting the longhorn engine for that node. This seemed to solved the issue, but after a while it came back.

We've tried to look into logs, but the only useful information we've found is in the kubelet:
```
I0417 09:13:26.365251   32725 reconciler.go:209] operationExecutor.VerifyControllerAttachedVolume started for volume "pvc-84d920a2-974e-4304-add6-5b31a4f81a3d" (UniqueName: "kubernetes.io/csi/driver.longhorn.io^pvc-84d920a2-974e-4304-add6-5b31a4f81a3d") pod "psql-test-vol-exp-postgresql-0" (UID: "2bc26ff0-9fe8-41b6-8fc0-cd2f06c39d03") 
E0417 09:13:26.369113   32725 nestedpendingoperations.go:270] Operation for "\"kubernetes.io/csi/driver.longhorn.io^pvc-84d920a2-974e-4304-add6-5b31a4f81a3d\"" failed. No retries permitted until 2020-04-17 09:15:28.36908213 +0000 UTC m=+84602.667818333 (durationBeforeRetry 2m2s). Error: "Volume not attached according to node status for volume \"pvc-84d920a2-974e-4304-add6-5b31a4f81a3d\" (UniqueName: \"kubernetes.io/csi/driver.longhorn.io^pvc-84d920a2-974e-4304-add6-5b31a4f81a3d\") pod \"psql-test-vol-exp-postgresql-0\" (UID: \"2bc26ff0-9fe8-41b6-8fc0-cd2f06c39d03\") "

```

This actually seems familiar to what faced in https://github.com/longhorn/longhorn/issues/267#issuecomment-444367116 but in that case it seems that a later kubernetes version solved the issue.

I'm sorry we don't have further logs to provide by honestly we've been looking pretty much everywhere and we cannot find any useful logs. You guys maybe have a hint about this. 

Thanks in advance.

**Comment by yasker** [comment]
@nicolamarella can you send us a support bundle at longhorn-support-bundle@rancher.com ? We will take a look.

And how many volumes you have per node? And how long it normally spend to make the volume usable for the pod?

**Comment by nicolamarella** [comment]
Hello @yasker, thank you for the reply, I've just send the bundle via email. 

At the moment we have around 15 to 20 volumes per node. On a normal situation it takes very little, i would say less than a minute or even quicker. But for some reason, when it gets stuck it can take up to 15 to 20 minutes, unless it fails completely and we need to manualy intervene.

**Comment by nicolamarella** [comment]
For those who have a similar issue, we found a temporary solution:
The problem appears to be originated when using `fsGroup` in the security context, and as you know, postgres chart uses it by default. We had many of them and also used in some custom charts. 

Disabling it and using an init-container to do chmod to fix the permission seems to have solved the (at least for now).

**Comment by yasker** [comment]
This might be related to #1154

**Comment by dgiebert** [comment]
Seems that restarting the kubelets helps for Longhorn to pick up the attachments, or they at least are showing up in the UI with a transition from detached to attached.

Checking the kubelet configuration in RKE could it be related to this configuration from the hardening guideline: `streaming-connection-idle-timeout: 1800s` ?

**Comment by ttpcodes** [comment]
Are you running into this issue on new `PersistentVolumes`, or just existing ones? If this issue is occurring on existing `Volumes`, and these `Volumes` hold a lot of data and are high in inodes, then I think this is connected to kubernetes/kubernetes#69699.

The permissions change to match `fsGroup` is not actually handled by `Longhorn`, but rather by `Kubernetes` by `kubelet` during the `Volume` attachment process. After `Longhorn` returns a response to the `NodePublishVolume` call, `kubelet` calls `SetVolumeOwnership` if `fsGroup` is set, and this operation can take a long time and time out on `Volumes` with many inodes to process.

Unfortunately, it doesn't look like there's much we can do to improve the processing time of `SetVolumeOwnership` on our end. The issue is being tracked at kubernetes/kubernetes#69699, and an option for skipping this ownership change under certain conditions was added in `1.18` (currently in `Alpha`): kubernetes/enhancements#695

**Comment by possani** [comment]
It seems that restarting the kube-apiservers makes it reattach the volumes. Of course, restarting Kubernetes components whenever we need a Volume to attach is not a solution.

@ttpcodes, it does not happen only to existing ones and many of the existing ones are quite empty.

**Comment by yasker** [comment]
We need to double-check if there is anything we can do from our end. If Longhorn is supporting e.g. MinIO which might have tens or hundreds of thousands of files, the attaching time can be hours or more.

**Comment by yasker** [comment]
Move to the icebox, pending on update from #2131

**Comment by innobead** [comment]
Already fixed in #2131 and https://longhorn.io/kb/troubleshooting-volume-take-long-time-to-mount/ provided.
